### ISSUE TITLE
Need an easy way to globally turn off all argument checks

### ISSUE THREAD

**Issue by mysticfall** [danielwertheim/Ensure.That#110]
I'm using this library quite extensively in a project which runs on top of a game engine. As the environment requires high performance, I wish there was an easy way to bypass all argument checks so that there wouldn't be unnecessary overhead in a release build.

I don't fully understand the reason behind removing of `Ensure.On/Off` feature, but if it was because of poor runtime performance, I think we still need an alternative to achieve what it did without incurring such an overhead.

**Comment by danielwertheim** [comment]
Haven't found a good way (not paying the price of the check) to bake that into the lib itself. So currently a construct for it would have to live in the consumer code. E.g. compiler directive or Conditional attribute, .... Maybe this could be solved using weaving with Fody?

**Comment by mysticfall** [comment]
Using compile time weaving like Fody would be an option, if it is supported on Mono which the game engine I'm using relies on.

I understand if there are some technical difficulties involved, and in that case, it'd be good if I can keep this issue open as a reminder so I could be notified when we have a viable solution in future.

**Comment by ndrwrbgs** [comment]
I’m working on Fody integration but it’ll take a while. There is a flag that this library uses to short circuit. Does Fody not work with Mono?

**Comment by ndrwrbgs** [comment]
Would you like to compile them out? It’s possible Daniel would accept a Conditional attribute addition to the library - but it’d only work on the method that don’t return anything (though I’m not sure if any of those exist anymore)

**Comment by ndrwrbgs** [comment]
It looks like the Ensure.That().Is(value) syntax is the only one that does not return a value, which is needed for the `ConditionalAttribute` to work. I believe you'll still pay for setting up the struct that is passed into those methods though in cases where JIT doesn't optimize that away (there's not ConditionalAttribute support for deep calls like that).

You could try to make that change in the library here and send for review, or it's always possible as @danielwertheim mentioned to wrap Ensure.That calls in your library through an intermediary layer that has On/Off logic or ConditionalAttribute's around the entire calls (thereby avoiding any possible Property accesses (e.g. Ensure.Collection is a property) and struct (Ensure.That) setups entirely.

TLDR there does not exist at this moment a direct solution for disabling yet.


BTW - Fody should work with Mono I believe. It uses Mono to do the IL modification. I presume the harder part would be getting it integrated with Unity or similar - I see this for some possible answers: https://github.com/Fody/Fody/issues/275

**Comment by danielwertheim** [comment]
I didn't mean that the `ConditionalAttribute` should lie in `Ensure.That`. We have tried that and failed. I meant in the code-base where Ensure.That is used or if that would be weaved somehow with Fody. Or instead of Condition, use compile time directives. I would accept Conditional in Ensure.That if there was a smooth way to get it to be opt out of E.T, but as far as I remember that was not possible:  #70

**Comment by danielwertheim** [comment]
This works... Maybe not as pretty, but it works and gives you one place to define the conditional in your code base.  The extra method will not be inlined if it's not built with optimized code, which is the default with  `DEBUG` builds, but I guess that doesn't matter for `DEBUG` builds, getting an extra call in the call stack that is.

```csharp
using System;
using System.Diagnostics;
using EnsureThat;
using static ConsoleApp2.Conditional;

namespace ConsoleApp2
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine($"{new MyThing("Test")}");

            Console.WriteLine($"{new MyThing(string.Empty)}");
        }
    }

    public class MyThing
    {
        public string Value { get; }

        public MyThing(string value)
        {
            IfDebug(() => EnsureArg.IsNotEmptyOrWhitespace(value, nameof(value)));

            Value = value;
        }

        public override string ToString() => Value;
    }

    internal static class Conditional
    {
        [Conditional("DEBUG")]
        internal static void IfDebug(Action a) => a();
    }
}
```

**Comment by ndrwrbgs** [comment]
@Daniel, that’s beautifully elegant in showing the intent of the code rather than hiding it in a conditional at the declaration side! I’m surprised I’ve never seen it before.

**Comment by danielwertheim** [comment]
@mysticfall the thing is that it need to exist in your code, dll's can not act on the project being compiled.

**Comment by mysticfall** [comment]
I know but what I suggested was to setup another project (and thus assembly) for the wrapper. Could you (or other contributors) consider doing it in future, if it would work that way?

**Comment by ndrwrbgs** [comment]
It doesn’t sound particularly difficult, but it would be something that had
to be maintained/kept up to date with the library.

As my team’s don’t effectively use debug builds (so debug only checks would
be the same as no checks at all) I don’t have personal investment in the
request.

**Comment by stfbln** [comment]
Hello, I have been using this library quite extensively in a multitude of project and indeed having a way to turn off all validations would be really nice. I am personally not a big fan of conditional flags and was wondering if creating "NoOp" instance would help. For example If `AnyArg` is changed to have virtual methods as such:

```
    public class AnyArg
    {
        [NotNull]
        [ContractAnnotation("value:null => halt")]
        public virtual T IsNotNull<T>([NoEnumeration, ValidatedNotNull] T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : class
        {
            if (value == null)
                throw Ensure.ExceptionFactory.ArgumentNullException(ExceptionMessages.Common_IsNotNull_Failed, paramName, optsFn);

            return value;
        }

        [NotNull]
        [ContractAnnotation("value:null => halt")]
        public virtual T? IsNotNull<T>([ValidatedNotNull] T? value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct
        {
            if (value == null)
                throw Ensure.ExceptionFactory.ArgumentNullException(ExceptionMessages.Common_IsNotNull_Failed, paramName, optsFn);

            return value;
        }

        public virtual T IsNotDefault<T>(T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct
        {
            if (default(T).Equals(value))
                throw Ensure.ExceptionFactory.ArgumentException(ExceptionMessages.ValueTypes_IsNotDefault_Failed, paramName, optsFn);

            return value;
        }
    }
```

And with a `NoOpAnyArg` class:

```
    public class NoOpAnyArg : AnyArg
    {
        [NotNull]
        [ContractAnnotation("value:null => halt")]
        public override T IsNotNull<T>([NoEnumeration, ValidatedNotNull] T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : class
        {
            return value;
        }

        [NotNull]
        [ContractAnnotation("value:null => halt")]
        public override T? IsNotNull<T>([ValidatedNotNull] T? value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct
        {
            return value;
        }

        public override T IsNotDefault<T>(T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct
        {
            return value;
        }
    }
```

Then the following could be done in `Ensure`:

```
public static class Ensure
{
        internal static readonly ExceptionFactory ExceptionFactory = new ExceptionFactory();

        public static void Off() {
               Any = new NoOpAnyArg();
        }

        /// <summary>
        /// Ensures for objects.
        /// </summary>
        [NotNull]
        public static AnyArg Any { get; private set; } = new AnyArg();

       // the rest of the class following the same pattern...
}
```
